File: common/src/main/java/org/bread_experts_group/breadlib/network/NetworkContext.java
package org.bread_experts_group.breadlib.network;

import net.minecraft.client.multiplayer.ClientLevel;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.level.Level;
import org.bread_experts_group.breadlib.platform.ApplicationSide;

public record NetworkContext(Player player) {
	public Level level() {
		return player.level();
	}

	public ServerLevel serverLevel() {
		if (this.getSide() != ApplicationSide.SERVER) throw new IllegalStateException("Current dist is not server.");
		return (ServerLevel) this.level();
	}

	public ClientLevel clientLevel() {
		if (this.getSide() != ApplicationSide.CLIENT) throw new IllegalStateException("Current dist is not client.");
		return (ClientLevel) this.level();
	}

	public ApplicationSide getSide() {
		return level().isClientSide ? ApplicationSide.CLIENT : ApplicationSide.SERVER;
	}
}

File: common/src/main/kotlin/org/bread_experts_group/breadlib/network/NetworkContext.kt
package org.bread_experts_group.breadlib.network

import net.minecraft.client.multiplayer.ClientLevel
import net.minecraft.server.level.ServerLevel
import net.minecraft.world.entity.player.Player
import net.minecraft.world.level.Level
import org.bread_experts_group.breadlib.platform.ApplicationSide

data class NetworkContext(val player: Player) {
	val side: ApplicationSide
		get() = if (level().isClientSide) ApplicationSide.CLIENT else ApplicationSide.SERVER

	fun level(): Level {
		return player.level()
	}

	fun serverLevel(): ServerLevel {
		check(this.side == ApplicationSide.SERVER) { "Current dist is not server." }
		return this.level() as ServerLevel
	}

	fun clientLevel(): ClientLevel {
		check(this.side == ApplicationSide.CLIENT) { "Current dist is not client." }
		return this.level() as ClientLevel
	}
}